kolibri/core/assets/src/mediator/mediator.js

```diff
@@ -1,3 +1,241 @@
-var Kolibri = {};
+'use strict';
+/**
+ * Mediator module.
+ * Provides the main controller and event dispatcher for the Kolibri FrontEnd core app.
+ * @module Mediator
+ */
 
-module.exports = Kolibri;
+
+var asset_loader = require('./asset_loader');
+var Backbone = require('backbone');
+var _ = require('lodash');
+var logging = require('loglevel');
+
+/**
+ * @constructor
+ * @classdesc The Mediator class - registers and loads plugins and acts as a global event dispatcher.
+ */
+var Mediator = function() {
+    /**
+     * Keep track of all registered plugins - object is of form:
+     * plugin_name: plugin_object
+     **/
+    this._plugin_registry = {};
+
+    /**
+     * Keep track of all callbacks that have been fired for as yet unloaded modules.
+     * plugin_name: {Function[]} of callbacks
+     **/
+    this._callback_buffer = {};
+
+    /**
+     * Keep track of all registered callbacks bound to events - this allows for easier stopListening later.
+     * plugin_name: {object} - event: {object} - method: callback function
+     **/
+    this._callback_registry = {};
+
+    /**
+     * Keep track of all registered async callbacks bound to events - this allows for easier stopListening later.
+     * plugin_name: {object[]} - with keys 'event' and 'callback'.
+     **/
+    this._async_callback_registry = {};
+
+    this._event_dispatcher = _.clone(Backbone.Events);
+};
+
+/**
+ * @param {Plugin} plugin - object of Plugin class
+ * @description Registers a plugin that has already been loaded into the frontend. Registers event listeners for
+ * multiple time and one time events. When all event listeners have been registered, any buffered callbacks are passed
+ * to the Plugin object, in case it was previously registered asynchronously.
+ */
+Mediator.prototype.register_plugin_sync = function(plugin) {
+
+    this._register_multiple_events(plugin);
+    this._register_one_time_events(plugin);
+
+    this._plugin_registry[plugin.name] = plugin;
+    this._clear_async_callbacks(plugin);
+    this._execute_callback_buffer(plugin);
```

Comment:
Memory leaks are a concern. Not too much of a problem so early on, but let's watch out for this in the future.